pango/src/layout.rs

```diff
@@ -0,0 +1,51 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use std::mem;
+
+use glib::translate::*;
+
+use crate::LayoutLine;
+
+/// The result of [`LayoutLine::x_to_index`].
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub struct HitPosition {
+    /// The UTF-8 byte offset of the grapheme closest to the position.
+    index: i32,
+    /// The codepoint within the grapheme of the position.
+    ///
+    /// This will always be either `0`, or the number of `char`s (*not bytes!*)
+    /// in the grapheme. This represents whether the user clicked near the start
+    /// of the grapheme or near the end; this is important for things like
+    /// resolving cursor positions.
+    trailing: i32,
+    /// Whether or not the position was within the bounds of the line.
+    ///
+    /// If this is `false`, then `index` and `trailing` will always resolve
+    /// to either the very first or the very last position in the line; this
+    /// behaviour is dependent on the line's resolved writing direction.
+    is_inside: bool,
+}
+
+
+impl LayoutLine {
+    /// Return the [`HitPosition`] corresponding to the provided `x_pos` in this line.
+    pub fn x_to_index(&self, x_pos: i32) -> HitPosition {
+        unsafe {
+            let mut index = mem::MaybeUninit::uninit();
```

Comment:
There isn't much interest into using `MaybeUninit` here.

Reply:
I just copy-pasted this from the existing impl. Would you prefer to just initialize to `0` and pass that?


This is a good reminder,  I wanted to document that the returned positions are relative to the first position in the line, not the larger layout.

Reply:
In the original code, we assumed that if it returned `false`, something went wrong and we just returned `None`. The situation changed since then. ;)

Reply:
Oh, I see what you're saying. 👍